File: advent-of-code/src/main/java/com/willmolloy/adventofcode/common/Input.java
package com.willmolloy.adventofcode.common;

import com.willmolloy.adventofcode.common.grid.CharGrid;
import java.util.List;

/**
 * Helper for reading the various inputs.
 *
 * @author <a href=https://willmolloy.com>Will Molloy</a>
 */
public interface Input {

  /** Read input lines. */
  List<String> lines();

  /** Read input num lines. */
  List<long[]> numLines();

  /** Read input string. */
  String string();

  /** Read input nums. */
  long[] nums();

  /** Read input {@link CharGrid}. */
  CharGrid charGrid();
}

File: advent-of-code/src/main/kotlin/com/willmolloy/adventofcode/common/Input.kt
package com.willmolloy.adventofcode.common

import com.willmolloy.adventofcode.common.grid.CharGrid

/**
 * Helps read the various advent-of-code inputs.
 *
 * @author <a href=https://willmolloy.com>Will Molloy</a>
 */
interface Input {

  /** Reads the input as a single string. */
  fun readString(): String

  /** Reads the input as a list of lines. */
  fun readLines(): List<String>

  /** Reads the input as space-separated numbers. */
  fun readNums(): LongArray

  /** Reads the input as lines of space-separated numbers. */
  fun readLinesOfNums(): List<LongArray>

  /** Reads the input as lines of digits. */
  fun readLinesOfDigits(): List<IntArray>

  /** Reads the input as a 2D character grid. */
  fun readCharGrid(): CharGrid
}
